dataloader/dataloader_impl.hpp

```diff
@@ -0,0 +1,147 @@
+  if (datasetMap.count(dataset))
+  {
+    // Use utility functions to download the dataset.
+    DownloadDataset(dataset);
+
+    if (datasetMap[dataset].loadCSV)
+    {
+      LoadCSV(datasetMap[dataset].trainPath, true, shuffle, ratio, useScaler,
+              datasetMap[dataset].dropHeader,
+              datasetMap[dataset].startTrainingInputFeatures,
+              datasetMap[dataset].endTrainingInputFeatures,
+              datasetMap[dataset].endTrainingPredictionFeatures,
+              datasetMap[dataset].endTrainingPredictionFeatures);
+
+      LoadCSV(datasetMap[dataset].testPath, false, false, useScaler,
+              datasetMap[dataset].dropHeader,
+              datasetMap[dataset].startTestingInputFeatures,
+              datasetMap[dataset].endTestingInputFeatures);
+    }
+
+    // Preprocess the dataset.
+    datasetMap[dataset].PreProcess(trainX, trainY, validX, validY, testX);
+  }
+  else
+  {
+    mlpack::Log::Fatal << "Unknown Dataset. " << dataset <<
+        " For other datasets try loading data using" <<
+        " generic dataloader functions such as LoadCSV." <<
+        " Refer documentation for more info." << std::endl;
+  }
+}
+
+
+template<
+  typename DatasetX,
+  typename DatasetY,
+  class ScalerType
+> void DataLoader<
+    DatasetX, DatasetY, ScalerType
+>::LoadCSV(const std::string &datasetPath,
+           const bool loadTrainData,
+           const bool shuffle,
+           const double ratio,
+           const bool useScaler,
+           const bool dropHeader,
+           const int startInputFeatures,
+           const int endInputFeatures,
+           const int startPredictionFeatures,
+           const int endPredictionFeatures,
+           const std::vector<std::string> augmentation,
+           const double augmentationProbability)
+{
+  arma::mat dataset;
+  data::Load(datasetPath, dataset, true);
+
+  dataset = dataset.submat(0, size_t(dropHeader), dataset.n_rows - 1,
+      dataset.n_cols - 1);
+
+  if (loadTrainData)
+  {
+    arma::mat trainDataset, validDataset;
+    data::Split(dataset, trainDataset, validDataset, ratio /* Add shuffle*/);
+
+    if (useScaler)
+    {
+      scaler.Fit(trainDataset);
+      scaler.Transform(trainDataset, trainDataset);
+      scaler.Transform(validDataset, validDataset);
+    }
+
+    trainX = trainDataset.rows(WrapIndex(startInputFeatures,
+        trainDataset.n_rows), WrapIndex(endInputFeatures,
+        trainDataset.n_rows));
+
+    trainY = trainDataset.rows(WrapIndex(startPredictionFeatures,
+        trainDataset.n_rows), WrapIndex(endPredictionFeatures,
+        trainDataset.n_rows));
+
+    validX = validDataset.rows(WrapIndex(startInputFeatures,
+        validDataset.n_rows), WrapIndex(endInputFeatures,
+        validDataset.n_rows));
+
+    validY = trainDataset.rows(WrapIndex(startPredictionFeatures,
+        validDataset.n_rows), WrapIndex(endPredictionFeatures,
+        validDataset.n_rows));
+
+    // Add support for augmentation here.
```

Comment:
We might want to add a TODO here, to make this easier to find.

Reply:
Will do.